packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -1,43 +1,258 @@
+        components: [
+          { name: 'success', type: 'bool' },
+          { name: 'result', type: 'bytes' },
+        ],
+      },
+    ],
+  },
+];
+
+export const supportedEVMChains: CaipChainId[] = [
+  'eip155:43114', // Avalanche
+  'eip155:8453', // Base
+  'eip155:1', // Ethereum
+];
+
+const TypeShape = M.or(...keys(YieldProtocol));
+const ChainShape = M.or(...supportedEVMChains);
+const PositionShape = M.splitRecord({}); // TODO
+
+const KeeperI = M.interface('keeper', {
+  add: M.call(
+    M.or(...keys(YieldProtocol)),
+    M.or(...values(PortfolioChain)),
+    M.remotable('OrchestrationAccount'),
+  ).returns(M.number()),
+  getPositions: M.call(TypeShape, ChainShape).returns(M.arrayOf(PositionShape)),
+  getAccount: M.call(M.number(), TypeShape).returns(
+    M.remotable('OrchestrationAccount'),
+  ),
+});
+
+const EvmTapI = M.interface('EvmTap', {
+  receiveUpcall: M.call(M.record()).returns(M.or(VowShape, M.undefined())),
+});
+
+type LocalAccount = OrchestrationAccount<{ chainId: 'agoric-any' }>;
 
-type Accounts = {
-  Aave: never; // TODO
-  Compound: never; // TODO
-  USDN: OrchestrationAccount<{ chainId: 'noble-any' }>;
+type EVMProtocolState = {
+  type: 'Aave' | 'Compound';
+  chain: CaipChainId;
+  localAccount: LocalAccount;
+  remoteAccountAddress?: string;
+  isActive: boolean;
+};
+
+/** ICA on Noble */
+type NobleAccount = OrchestrationAccount<{ chainId: 'noble-any' }>;
+type NobleDollarState = {
+  type: 'USDN';
+  chain: CaipChainId; // cosmos:noble-...
+  account: NobleAccount;
+  isActive: boolean;
+};
+
+type AccountOf = {
+  Aave: LocalAccount;
+  Compound: LocalAccount;
+  USDN: NobleAccount;
+};
+
+type PositionInfo = EVMProtocolState | NobleDollarState;
+type PortfolioKitState = {
+  positions: MapStore<number, PositionInfo>;
 };
 
 export const preparePortfolioKit = (zone: Zone) =>
   zone.exoClassKit(
     'Portfolio',
-    interfaceTODO,
-    () =>
-      ({
-        USDN: undefined,
-      }) as Partial<Accounts>,
     {
+      keeper: KeeperI,
+      invitationMakers: M.interface('invitationMakers', {
+        // TODO
+        // supplyToAave: M.call(M.record()).returns(M.promise()),
+        // borrowFromAave: M.call(M.record()).returns(M.promise()),
+        // supplyToCompound: M.call(M.record()).returns(M.promise()),
+        // borrowFromCompound: M.call(M.record()).returns(M.promise()),
+        // withdraw: M.call(M.record()).returns(M.promise()),
+      }),
+      tap: EvmTapI,
+    },
+
+    (): PortfolioKitState => ({ positions: zone.mapStore('positions') }),
+    {
+      tap: {
+        receiveUpcall(event: VTransferIBCEvent) {
+          trace('receiveUpcall', event);
+
+          const tx: FungibleTokenPacketData = JSON.parse(
+            atob(event.packet.data),
+          );
+
+          trace('receiveUpcall packet data', tx);
+          const memo: AxelarGmpIncomingMemo = JSON.parse(tx.memo);
+
+          if ((supportedEVMChains as string[]).includes(memo.source_chain)) {
```

Comment:
to reduce the level of indentation, I wonder about using returns:

```suggestion
          if (!(supportedEVMChains as string[]).includes(memo.source_chain)) {
            return;
          }
```